drivers/raw_exec/driver.go

```diff
@@ -0,0 +1,530 @@
+		pid:          taskState.Pid,
+		pluginClient: pluginClient,
+		task:         taskState.TaskConfig,
+		procState:    drivers.TaskStateRunning,
+		startedAt:    taskState.StartedAt,
+		exitResult:   &drivers.ExitResult{},
+	}
+
+	r.tasks.Set(taskState.TaskConfig.ID, h)
+
+	go h.run()
+	return nil
+}
+
+func (r *RawExecDriver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, error) {
+	if _, ok := r.tasks.Get(cfg.ID); ok {
+		return nil, fmt.Errorf("task with ID '%s' already started", cfg.ID)
+	}
+
+	var driverConfig TaskConfig
+	if err := cfg.DecodeDriverConfig(&driverConfig); err != nil {
+		return nil, err
+	}
+
+	handle := drivers.NewTaskHandle(pluginName)
+	handle.Config = cfg
+
+	pluginLogFile := filepath.Join(cfg.TaskDir().Dir, "executor.out")
+	executorConfig := &dstructs.ExecutorConfig{
+		LogFile:  pluginLogFile,
+		LogLevel: "debug",
+	}
+
+	exec, pluginClient, err := utils.CreateExecutor(os.Stderr, hclog.Debug, 14000, 14512, executorConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	execCmd := &executor.ExecCommand{
+		Cmd:                driverConfig.Command,
+		Args:               driverConfig.Args,
+		Env:                cfg.EnvList(),
+		User:               cfg.User,
+		BasicProcessCgroup: !r.config.NoCgroups,
+		TaskDir:            cfg.TaskDir().Dir,
+		StdoutPath:         cfg.StdoutPath,
+		StderrPath:         cfg.StderrPath,
+	}
+
+	ps, err := exec.Launch(execCmd)
+	if err != nil {
+		pluginClient.Kill()
+		return nil, err
+	}
+
+	h := &rawExecTaskHandle{
+		exec:         exec,
+		pid:          ps.Pid,
+		pluginClient: pluginClient,
+		task:         cfg,
+		procState:    drivers.TaskStateRunning,
+		startedAt:    time.Now().Round(time.Millisecond),
+		logger:       r.logger,
+	}
+
+	r.tasks.Set(cfg.ID, h)
+
+	driverState := RawExecTaskState{
+		ReattachConfig: utils.ReattachConfigFromGoPlugin(pluginClient.ReattachConfig()),
+		Pid:            ps.Pid,
+		TaskConfig:     cfg,
+		StartedAt:      h.startedAt,
+	}
+
+	if err := handle.SetDriverState(&driverState); err != nil {
+		r.logger.Error("failed to start task, error setting driver state", "error", err)
+		exec.Shutdown("", 0)
+		pluginClient.Kill()
+		return nil, err
+	}
+
+	go h.run()
+	return handle, nil
+}
+
+func (r *RawExecDriver) WaitTask(ctx context.Context, taskID string) (<-chan *drivers.ExitResult, error) {
+	ch := make(chan *drivers.ExitResult)
+	handle, ok := r.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+	go r.handleWait(ctx, handle, ch)
+
+	return ch, nil
+}
+
+func (r *RawExecDriver) handleWait(ctx context.Context, handle *rawExecTaskHandle, ch chan *drivers.ExitResult) {
+	defer close(ch)
+	var result *drivers.ExitResult
+	ps, err := handle.exec.Wait()
+	if err != nil {
+		result = &drivers.ExitResult{
+			Err: fmt.Errorf("executor: error waiting on process: %v", err),
+		}
+	} else {
+		result = &drivers.ExitResult{
+			ExitCode: ps.ExitCode,
+			Signal:   ps.Signal,
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		return
```

Comment:
Hmm yeah, ideally canceling the context would unblock the exec.Wait call. Again the caller is really the GRPC server:

https://github.com/hashicorp/nomad/blob/f-driver-plugin-raw_exec/plugins/drivers/server.go#L112

It might be better to wrap that in a select with the ctx.Done() instead of doing it in the driver implementation.